server/src/embedded_support.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+
+/**
+ * Takes a TS file and strips out all non-inline template content.
+ *
+ * This process is the same as what's done in the VSCode example for embedded languages.
+ *
+ * Note that the example below implements the support on the client side. This is done on the server
+ * side to enable language services in other editors to take advantage of this feature by depending
+ * on the @angular/language-server package.
+ *
+ * @see https://github.com/microsoft/vscode-extension-samples/blob/fdd3bb95ce8e38ffe58fc9158797239fdf5017f1/lsp-embedded-request-forwarding/client/src/embeddedSupport.ts#L131-L141
+ * @see https://code.visualstudio.com/api/language-extensions/embedded-languages
+ */
+export function getHTMLVirtualContent(documentText: string): string {
+  const sf =
+      ts.createSourceFile('temp', documentText, ts.ScriptTarget.ESNext, true /* setParentNodes */);
+  const inlineTemplateNodes: ts.Node[] = findAllMatchingNodes(sf, isInlineTemplateNode);
+
+  // Create a blank document with same text length
+  let content = documentText.split('\n')
+                    .map(line => {
+                      return ' '.repeat(line.length);
+                    })
+                    .join('\n');
```

Comment:
In the real-world extension, the vscode checks the line break by the symbol `\n` and `\r`.

https://github.com/microsoft/vscode/blob/f4d9b1d53db3118f3be5e5efaf83affab670502b/extensions/html-language-features/server/src/modes/embeddedSupport.ts#L198-L214